src/main/java/org/truffleruby/cext/CExtNodes.java

```diff
@@ -1040,34 +1041,66 @@ public DynamicObject classNew(VirtualFrame frame, DynamicObject superclass) {
     @CoreMethod(names = "rb_tr_debug", onSingleton = true, rest = true)
     public abstract static class DebugNode extends CoreMethodArrayArgumentsNode {
 
-        @TruffleBoundary
+        @Child CallDispatchHeadNode toSCall;
+
         @Specialization
-        public Object debug(Object... objects) {
+        public Object debug(VirtualFrame frame, Object... objects) {
             if (objects.length > 1) {
-                System.err.printf("Printing %d values%n", objects.length);
+                printf("Printing %d values%n", objects.length);
             }
             for (Object object : objects) {
                 final String representation;
-
                 if (RubyGuards.isRubyString(object)) {
-                    final Rope rope = StringOperations.rope((DynamicObject) object);
-                    final String str = RopeOperations.decodeRope(rope);
-                    final byte[] bytes = rope.getBytes();
-                    final StringBuilder builder = new StringBuilder();
-                    for (int i = 0; i < bytes.length; i++) {
-                        if (i % 4 == 0 && i != 0 && i != bytes.length-1) {
-                            builder.append(" ");
-                        }
-                        builder.append(String.format("%02x", bytes[i]));
-                    }
-                    representation = str + " (" + builder.toString() + ")";
+                    representation = getRubyStringRepresentation((DynamicObject) object);
+                } else if (RubyGuards.isRubyBasicObject(object)) {
+                    representation = getString(object) + " (" +
+                            StringOperations.getString(callToS(frame, object)) + ")";
                 } else {
-                    representation = object.toString();
+                    representation = getString(object);
                 }
-                System.err.printf("%s @ %s: %s%n", object.getClass(), System.identityHashCode(object), representation);
+                printf("%s @ %s: %s%n", object.getClass(), identityHashCode(object), representation);
             }
             return nil();
         }
+
+        private DynamicObject callToS(VirtualFrame frame, Object object) {
+            if (toSCall == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                toSCall = CallDispatchHeadNode.create();
+            }
+
+            return (DynamicObject) toSCall.call(frame, object, "to_s");
+        }
```

Comment:
You can avoid needing the frame here if you use `CallDispatchHeadNode.createOnSelf()` (ignores visibility), and then you can pass `null` safely instead of frame.
I'll look at having javadoc and maybe a different interface for private calls so this is clearer the frame is not needed in that case.
The advantage is there is no need to add all these fine-grained boundaries, and it doesn't make `debug()` runtime-compilable, which is not interesting and increases the SVM image size needlessly.

Reply:
done

Reply:
I still see the old code with a frame.